fix(core): update unknown tag error for jit standalone components

dario-piotrowicz opened the pull request:
update the error message presented during jit compilation when an unrecognized
tag/element is found in a standalone component so that it does not mention
the ngModule anymore

Note: the aot variant is present in PR #45919

resolves #45818

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [x] Feature (may be a bugfix)
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## Issue

Issue Number: #45818


## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


## Other information

Result:
![JIT](https://user-images.githubusercontent.com/61631103/167289309-4dbf2835-6609-45bc-b712-8824dfc844c2.png)

dario-piotrowicz commented on packages/core/src/render3/instructions/lview_debug.ts:
```diff
@@ -160,7 +160,7 @@ export const TViewConstructor = class TView implements ITView {
       public _decls: number,
       public _vars: number,
-
+      public standalone: boolean,
```
adding the standalone flag to the `Tview` seemed to me the cleanest way to get that info in the `validateElementIsKnown`, I hope that makes sense :sweat:

dario-piotrowicz commented:
I wanted to add unit tests for this, but it does not seem like there are any unit tests regarding this sort of functionality, please let me know if I should add some somewhere

(I may be wrong but I imagine that jit is almost deprecated in favour of aot? so maybe that's why unit testing functionalities here is not as important?)

AndrewKushnir commented:
@dario-piotrowicz thanks for creating this PR! I think we should try to use the approach proposed by Joost (in https://github.com/angular/angular/issues/45818#issuecomment-1120960917) vs adding it to the `TView`. Since this information is not needed anywhere other than the error message, we'd be able to tree-shake away that code in prod bundles (unlike the changes in `TView`, which would not be tree-shakable). 

Quick comment: let's update the commit message to say `fix(core): ...`, since this is the fix to the error message, which looks incorrect in a context of standalone components (the second PR can be marked as a fix as well).

pkozlowski-opensource commented:
@dario-piotrowicz yes, I still believe that we should add relevant tests for the JIT case to exercise changes to packages/core/src/render3/instructions/element.ts. As for the test location I've suggested on in my previous comment  https://github.com/angular/angular/issues/45818#issuecomment-1120950658

dario-piotrowicz commented:
@AndrewKushnir @pkozlowski-opensource Thank you very much :slightly_smiling_face: 

I'll address all your comments this coming weekend, sorry that it's been a slow few days :sweat:

dario-piotrowicz committed:
fix(core): update unknown tag error for jit standalone components

update the error message presented during jit compilation when an unrecognized
tag/element is found in a standalone component so that it does not mention
the ngModule anymore

Note: the aot variant is present in PR #45919

resolves #45818

dario-piotrowicz commented on packages/core/test/acceptance/standalone_spec.ts:
```diff
@@ -628,6 +628,68 @@ describe('standalone components, directives and pipes', () => {
+          expect(spy).toHaveBeenCalledWith(jasmine.stringMatching(errorRegexB));
+        });
+
+        it('should not warn the user when an unknown element is present inside an ng-template',
```
@pkozlowski-opensource is this what you had in mind in https://github.com/angular/angular/issues/45818#issuecomment-1120950658? :slightly_smiling_face:

dario-piotrowicz replied:
I've added the test with the `[ngIf]="true]` below :slightly_smiling_face:

AndrewKushnir commented on packages/core/src/render3/instructions/element.ts:
```diff
@@ -54,9 +56,17 @@ function elementStartFirstCreatePass(
+    const hasDirectives =
+        resolveDirectives(tView, lView, tNode, getConstant<string[]>(tViewConsts, localRefsIndex));
+
+    const declarationLView = lView[DECLARATION_COMPONENT_VIEW] as LView<Type<unknown>>;
+    const context = declarationLView[CONTEXT];
+    const def = getComponentDef(context.constructor);
+    const hostIsStandalone = !!(def?.standalone);
```
I feel like we may need this logic in other error messages as we start to update them with standalone context more. I'd like to propose that we extract it to a separate function (for ex. `isHostComponentStandalone` and put it in `render3/instructions/shared.ts` file) and add a check like this:
https://github.com/angular/angular/blob/ca04ecedec42bc15a8135936ae33a8850cbd641d/packages/core/src/render3/state.ts#L355
It should help avoid adding this relatively slow path into the perf-sensitive code (like change detection). If we come across a situation when we need similar detection in the non-dev mode, we'd need to re-consider adding a flag to the `TView` (as it was implemented originally in this PR).

dario-piotrowicz replied:
sounds good! :slightly_smiling_face::+1:

I actually did think of creating a function for this piece of code but I was afraid of overengineer things :sweat_smile:

pkozlowski-opensource replied:
Agree with @AndrewKushnir - we should have a utility function here.

dario-piotrowicz replied:
I've added the `isHostComponentStandalone` function please have a look :slightly_smiling_face: 

(by the way, seems like the other functions in the shared file don't have unit tests associated to them, what do you think should I add one for the new function? if so could you suggest a file? :pray:)

___

Also, would it be worth it to also extract this two lines into another function?
```ts
  const declarationLView = lView[DECLARATION_COMPONENT_VIEW] as LView<Type<unknown>>;
  const context = declarationLView[CONTEXT];
```
(something like `getLViewContext` or something along those lines), let me know if this is a good idea and in that case if you'd like me to do it here or in a separate PR :slightly_smiling_face:

dario-piotrowicz replied:
@AndrewKushnir , @pkozlowski-opensource any comment on this one? :slightly_smiling_face:

AndrewKushnir commented on packages/core/src/render3/instructions/element.ts:
```diff
@@ -247,15 +258,18 @@ function validateElementIsKnown(
-          tagName}' is an Angular component, then verify that it is part of this module.\n`;
+      message += `1. If '${tagName}' is an Angular component, then verify that it is ${
+          hostIsStandalone ? 'included in the \'@Component.imports\' of this component' :
+                             'part of this module'}.\n`;
```
```suggestion
                             'a part of this module'}.\n`;
```

pkozlowski-opensource commented on packages/core/test/acceptance/standalone_spec.ts:
```diff
@@ -628,6 +628,87 @@ describe('standalone components, directives and pipes', () => {
     });
+
+    describe('templates without NO_ERRORS_SCHEMA', () => {
+      describe('unknown elements', () => {
```
nit: do we need the nested `describe` blocks here? IMO one block would be enough. Alternatively pull it into the existing `describe('schemas', () => {` block.

dario-piotrowicz replied:
ah yeah good point, I am not sure why I did it this way :sweat: 

anyways I removed the inner describe block and changed the text to `unknown template elements` that should be clear enough I hope, what do you think is that ok? :slightly_smiling_face: 

(by the way, I can put this new describe inside the schemas one, I didn't do that as here nothing is being directly done with schemas, so I am not sure I am totally a fan of that, but if you'd prefer it that way I can move it there :slightly_smiling_face: )

pkozlowski-opensource commented on packages/core/src/render3/instructions/element.ts:
```diff
@@ -54,9 +56,17 @@ function elementStartFirstCreatePass(
+
+    const declarationLView = lView[DECLARATION_COMPONENT_VIEW] as LView<Type<unknown>>;
+    const context = declarationLView[CONTEXT];
+    const def = getComponentDef(context.constructor);
```
So I'm nervous about the usage of the `context.constructor` here as it is not 100% reliable AFAIK. Kind of thinking that maybe we should store the `isStandalone` flag on the `TView` instead.

dario-piotrowicz replied:
I've added a `isHostComponentStandalone` function as suggested by Andew, I hope that's fine :slightly_smiling_face: 

if not please let me know :slightly_smiling_face:

dario-piotrowicz committed:
fixup! fix(core): update unknown tag error for jit standalone components

add isHostComponentStandalone

AndrewKushnir commented on packages/core/src/render3/instructions/element.ts:
```diff
@@ -54,9 +54,14 @@ function elementStartFirstCreatePass(
-      resolveDirectives(tView, lView, tNode, getConstant<string[]>(tViewConsts, localRefsIndex));
-  ngDevMode && validateElementIsKnown(native, tNode.value, tView.schemas, hasDirectives);
+  if (ngDevMode) {
+    const hasDirectives =
+        resolveDirectives(tView, lView, tNode, getConstant<string[]>(tViewConsts, localRefsIndex));
+
+    const hostIsStandalone = isHostComponentStandalone(lView);
+
+    validateElementIsKnown(native, tNode.value, tView.schemas, hasDirectives, hostIsStandalone);
+  }
```
super nit: I think we can remove new lines in between:
```suggestion
    const hasDirectives =
        resolveDirectives(tView, lView, tNode, getConstant<string[]>(tViewConsts, localRefsIndex));
    const hostIsStandalone = isHostComponentStandalone(lView);
    validateElementIsKnown(native, tNode.value, tView.schemas, hasDirectives, hostIsStandalone);
  }
```
Usually I use new lines to separate a particular block of code from the rest of the code, but in our case this is a single logical block, so we can just remove line breaks.

AndrewKushnir commented on packages/core/src/render3/instructions/shared.ts:
```diff
@@ -266,6 +268,20 @@ export function createTNodeAtIndex(
 }
 
+/**
+ * Checks if the component of a given view is a standalone one or not.
```
```suggestion
 * Checks if the current component is declared inside of a standalone component template.
```

AndrewKushnir commented on packages/core/src/render3/instructions/shared.ts:
```diff
@@ -266,6 +268,20 @@ export function createTNodeAtIndex(
+/**
+ * Checks if the component of a given view is a standalone one or not.
+ *
+ * @param lView The `LView` containing the view's data
+ * @returns `true` if the component owning this view is a standalone one, `false` otherwise
```
I'd suggest that we drop `@returns` as it's covered in the function description.
```suggestion
 * @param lView An `LView` that represents a current component that is being rendered.
```

AndrewKushnir commented on packages/core/src/render3/instructions/shared.ts:
```diff
@@ -266,6 +268,20 @@ export function createTNodeAtIndex(
+
+  const declarationLView = lView[DECLARATION_COMPONENT_VIEW] as LView<Type<unknown>>;
+  const context = declarationLView[CONTEXT];
+  const def = getComponentDef(context.constructor);
+  return !!(def?.standalone);
```
```suggestion
  const componentDef = getComponentDef(context.constructor);
  return !!(componentDef?.standalone);
```

AndrewKushnir reviewed (commented):
@dario-piotrowicz thanks for addressing the feedback 👍
I've left some final comments from my side (Pawel may provide some additional feedback as well).

dario-piotrowicz committed:
fixup! fix(core): update unknown tag error for jit standalone components

AndrewKushnir reviewed (approved):
👍

AndrewKushnir commented:
[Presubmit](http://test/OCL:449597217:BASE:449597243:1652914246287:67afc63a).

pkozlowski-opensource reviewed (approved):
LGTM

AndrewKushnir commented on packages/core/src/render3/instructions/element.ts:
```diff
@@ -54,9 +54,12 @@ function elementStartFirstCreatePass(
-  const hasDirectives =
-      resolveDirectives(tView, lView, tNode, getConstant<string[]>(tViewConsts, localRefsIndex));
-  ngDevMode && validateElementIsKnown(native, tNode.value, tView.schemas, hasDirectives);
+  if (ngDevMode) {
+    const hasDirectives =
+        resolveDirectives(tView, lView, tNode, getConstant<string[]>(tViewConsts, localRefsIndex));
+    const hostIsStandalone = isHostComponentStandalone(lView);
```
```suggestion
    const hasDirectives =
        resolveDirectives(tView, lView, tNode, getConstant<string[]>(tViewConsts, localRefsIndex));
    if (ngDevMode) {
        const hostIsStandalone = isHostComponentStandalone(lView);
```

AndrewKushnir replied:
@dario-piotrowicz we've ran into some issues during the presubmit and it turned out that the origin of it is the fact that the `resolveDirectives` got moved within the `ngDevMode` block, so it's removed in prod builds. However the line is load-bearing for prod mode as well. Could you please update it?

Also coincidentally the CI didn't start for the PR, so we were unable the verify the code during the CI run (it's very likely that the CI would've caught it).

dario-piotrowicz replied:
sure :slightly_smiling_face::+1:

dario-piotrowicz replied:
done :slightly_smiling_face:, sorry for the inconvenience :sweat:

dario-piotrowicz committed:
fixup! fix(core): update unknown tag error for jit standalone components

move resolveDirectives outside ngDevMode block

AndrewKushnir reviewed (approved):
@dario-piotrowicz thanks for addressing the feedback 👍 I'm starting a new presubmit and will add the PR to the merge queue if there are no failing targets.

AndrewKushnir commented:
[Presubmit](http://test/OCL:449846345:BASE:449846380:1652999411972:1cee7b50).

dario-piotrowicz committed:
fixup! fix(core): update unknown tag error for jit standalone components

remove debugger

dario-piotrowicz committed:
fixup! fix(core): update unknown tag error for jit standalone components

add missing semicolon

AndrewKushnir commented on packages/core/src/render3/instructions/shared.ts:
```diff
@@ -266,6 +268,19 @@ export function createTNodeAtIndex(
+  !ngDevMode && throwError('Must never be called in production mode');
+
+  const declarationLView = lView[DECLARATION_COMPONENT_VIEW] as LView<Type<unknown>>;
+  const context = declarationLView[CONTEXT];
+  const componentDef = getComponentDef(context.constructor);
+  return !!(componentDef?.standalone);
```
@dario-piotrowicz it looks like some tests are failing on CI because the `context` becomes undefined in some cases. Looking at some of the tests in the `render3` folder, they seem to use old testing infrastructure and we didn't have a chance to convert them to the new one yet. However, according to the docs for the `lView[CONTEXT]` field:
https://github.com/angular/angular/blob/0e14df697ad00a094b3a2a07e6e5e4295e978830/packages/core/src/render3/interfaces/view.ts#L176-L183
we should account for the fact that the `lView[CONTEXT]` might be `null` or `undefined`.

So I'd propose adding a check like this:
```suggestion
  const context = declarationLView[CONTEXT];
  
  // Unable to obtain a context, fall back to the non-standalone scenario.
  if (!context) return false;
  
  const componentDef = getComponentDef(context.constructor);
  return !!(componentDef?.standalone);
```

dario-piotrowicz replied:
sounds good, thanks a lot! :slightly_smiling_face:

dario-piotrowicz committed:
fixup! fix(core): update unknown tag error for jit standalone components

add no context check

AndrewKushnir commented:
[Presubmit](http://test/OCL:449846345:BASE:450102044:1653096249424:2b9430b0).
